internal/controller/appconfigurationprovider_controller.go

```diff
@@ -379,28 +467,41 @@ func newProviderStatus(phase acpv1.AppConfigurationSyncPhase, message string, sy
 	}
 }
 
-func (reconciler *AzureAppConfigurationProviderReconciler) shouldReconcile(provider *acpv1.AzureAppConfigurationProvider, existingConfigMap *corev1.ConfigMap, existingSecret *corev1.Secret) bool {
+func (reconciler *AzureAppConfigurationProviderReconciler) shouldReconcile(
+	provider *acpv1.AzureAppConfigurationProvider,
+	existingConfigMap *corev1.ConfigMap,
+	existingSecrets map[string]corev1.Secret) bool {
 	// Get the name and namespace of the provider
 	namespacedName := types.NamespacedName{
 		Name:      provider.Name,
 		Namespace: provider.Namespace,
 	}
+
 	if provider.Generation != reconciler.ProvidersReconcileState[namespacedName].Generation {
 		// If the provider is updated, we need to reconcile anyway
 		return true
 	}
+
 	if reconciler.ProvidersReconcileState[namespacedName].ConfigMapResourceVersion == nil ||
 		*reconciler.ProvidersReconcileState[namespacedName].ConfigMapResourceVersion != existingConfigMap.ResourceVersion {
 		// If the ConfigMap is removed or updated, we need to reconcile anyway
 		return true
 	}
-	if provider.Spec.Secret != nil &&
-		(reconciler.ProvidersReconcileState[namespacedName].SecretResourceVersion == nil ||
-			*reconciler.ProvidersReconcileState[namespacedName].SecretResourceVersion != existingSecret.ResourceVersion) {
-		// If the Secret is removed or updated, we need to reconcile anyway
+
+	if provider.Spec.Secret == nil {
+		return false
+	}
+
+	if len(reconciler.ProvidersReconcileState[namespacedName].ExistingSecretReferences) == 0 {
 		return true
 	}
 
+	for name, secret := range existingSecrets {
+		if reconciler.ProvidersReconcileState[namespacedName].ExistingSecretReferences[name].SecretResourceVersion != secret.ResourceVersion {
+			return true
+		}
+	}
+
```

Comment:
Is it possible that the loop of updating secret in `createOrUpdateSecrets` breaks early? ending up with a state that `len(reconciler.ProvidersReconcileState[namespacedName].ExistingSecretReferences) == 0` is true and `ExistingSecretReferences[name].SecretResourceVersion == secret.ResourceVersion` is true for all existing secrets, but the ir lengths don't match the length of `settings.SecretSettings` hence needs reconcile.

Reply:
Updated. And moved this `shouldReconcile()` function to `processor` to make the detailed logic transparency to the controller.